cosmoz-page-location.html

```diff
@@ -1,238 +1,173 @@
 					},
 
-					/**
-					 */
-					routeQueryString: {
-						type: String
+					routePath: {
+						type: String,
+						notify: true,
+						observer: '_routeChanged'
 					},
 
-					/**
-					 *
-					 */
-					routeQueryParams: {
+					routeQuery: {
 						type: Object,
+						value: defaultObject,
 						notify: true
 					},
 
-					/**
-					 */
 					routeHash: {
-						type: String,
-						notify: true
-					},
-
-					routeHashParams: {
 						type: Object,
+						value: defaultObject,
 						notify: true
 					},
 
-					_urlQueryString: {
-						type: String
-					},
-
-					_urlHash: {
-						type: String
-					},
-
-					_tempParamsObject: {
-						type: Object
-					},
-
-					_tempParamsString: {
-						type: String
+					hashBang: {
+						type: Boolean,
+						readOnly: true
 					}
 				},
 
 				observers: [
-					'_computeRoutePathAndParams(_urlHash)',
-					'_updateHash(routePath, routeQueryString, routeHash)'
+					'_routeChanged(routeQuery.*)',
+					'_routeChanged(routeHash.*)',
+					'_appHashChanged(_appHashString)',
 				],
 
-				attached: function () {
-					this._attached = true;
+				_paramsToString(value) {
+					if (typeof value === 'string') {
+						return value;
+					}
+					return this.$.params._encodeParams(value)
+						.replace(/%3F/g, '?').replace(/%2F/g, '/').replace(/'/g, '%27');
 				},
 
-				getRouteUrl: function (routeParams, routeHashParams) {
-					var prefix = this._hashBang ? '#!' : '#';
-
-					return prefix + this._getRoute(routeParams, routeHashParams);
+				_paramsFromString(value) {
+					return this.$.params._decodeParams(value);
 				},
 
-				getAppUrl: function (appParams, routeParams, routeHashParams) {
-					var route = this._getRoute(routeParams, routeHashParams),
-						queryString;
+				_microUpdate(path, value = {}) {
+					const keys = Object.keys,
+						current = this.get(path),
+						currentKeys = keys(current);
+					let valueKeys = keys(value);
 
-					this._tempParamsObject = Object.assign({}, this.appQueryParams, appParams);
-					queryString = this._tempParamsString;
-
-					return this._getUrl(this.appPath, queryString, route);
+					currentKeys
+						.filter(key => valueKeys.indexOf(key) < 0)
+						.forEach(key => value[key] = null);
 
+					keys(value)
+						.filter(key => current[key] !== value[key])
+						.forEach(key => this.set([path, key], value[key]));
 				},
 
-				_getRoute: function (routeParams, routeHashParams) {
-					var queryString,
-						hash;
-
-					this._tempParamsObject = Object.assign({}, this.routeQueryParams, routeParams);
-					queryString = this._tempParamsString;
-					this._tempParamsObject = Object.assign({}, this.routeHashParams, routeHashParams);
-					hash = this._tempParamsString;
-					return this._getUrl(this.routePath, queryString, hash);
+				_parse(appHashString) {
+					let regexp = /(!(?=\/))?([^#?]+)?#?([^?]+)?\??(.+)?$/gi,
+						matches = regexp.exec(appHashString);
+					matches = matches && matches.slice(1);
```

Comment:
What about `if (!matches)` ? Should be an return/exit or the return below will fail when referencing indexes of null.
`slice()` shouldn't be necessary, rather increase the indexes below? Enables `matches` to be `const`.